File: reader_writer(semaphore).c
#include <pthread.h>
#include <semaphore.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

// Maximum number of operations for readers and writers
#define MAX_OPERATIONS 20

sem_t rw_mutex;      // Semaphore to ensure mutual exclusion for writers
sem_t mutex;         // Semaphore to protect the read_count variable
int read_count = 0;  // Number of readers currently accessing the shared resource

// Shared resource
int shared_data = 0;
int total_operations = 0; // Tracks total read/write operations
pthread_mutex_t op_mutex = PTHREAD_MUTEX_INITIALIZER; // Protects total_operations

// Writer function
void* writer(void* arg) {
    int writer_id = ((int)arg);

    while (1) {
        pthread_mutex_lock(&op_mutex);
        if (total_operations >= MAX_OPERATIONS) {
            pthread_mutex_unlock(&op_mutex);
            break;
        }
        total_operations++;
        pthread_mutex_unlock(&op_mutex);

        sem_wait(&rw_mutex); // Lock the resource for writing

        // Writing to the shared resource
        shared_data += 1;
        printf("Writer %d wrote: %d\n", writer_id, shared_data);

        sem_post(&rw_mutex); // Release the resource
        sleep(rand() % 2 + 1); // Simulate writing time
    }

    return NULL;
}

// Reader function
void* reader(void* arg) {
    int reader_id = ((int)arg);

    while (1) {
        pthread_mutex_lock(&op_mutex);
        if (total_operations >= MAX_OPERATIONS) {
            pthread_mutex_unlock(&op_mutex);
            break;
        }
        total_operations++;
        pthread_mutex_unlock(&op_mutex);

        sem_wait(&mutex); // Lock to update read_count
        read_count++;
        if (read_count == 1) {
            sem_wait(&rw_mutex); // First reader locks the resource
        }
        sem_post(&mutex); // Release the lock on read_count

        // Reading the shared resource
        printf("Reader %d read: %d\n", reader_id, shared_data);

        sem_wait(&mutex); // Lock to update read_count
        read_count--;
        if (read_count == 0) {
            sem_post(&rw_mutex); // Last reader releases the resource
        }
        sem_post(&mutex); // Release the lock on read_count

        sleep(rand() % 2 + 1); // Simulate reading time
    }

    return NULL;
}

// Main function
int main() {
    pthread_t readers[3], writers[2]; // Array of reader and writer threads
    int reader_ids[3] = {1, 2, 3};   // Reader IDs
    int writer_ids[2] = {1, 2};      // Writer IDs

    // Initialize semaphores
    sem_init(&rw_mutex, 0, 1); // Binary semaphore for rw_mutex
    sem_init(&mutex, 0, 1);    // Binary semaphore for mutex

    // Create writer threads
    for (int i = 0; i < 2; i++) {
        pthread_create(&writers[i], NULL, writer, &writer_ids[i]);
    }

    // Create reader threads
    for (int i = 0; i < 3; i++) {
        pthread_create(&readers[i], NULL, reader, &reader_ids[i]);
    }

    // Join threads
    for (int i = 0; i < 2; i++) {
        pthread_join(writers[i], NULL);
    }
    for (int i = 0; i < 3; i++) {
        pthread_join(readers[i], NULL);
    }

    // Destroy semaphores (cleanup)
    sem_destroy(&rw_mutex);
    sem_destroy(&mutex);

    printf("All operations completed. Exiting.\n");

    return 0;
}

File: reader_writer(semaphore).py
import threading
import time
import random

# Shared resource (initialized to None to indicate no data available)
shared_resource = None

# Semaphores
mutex = threading.Semaphore(1)         # To protect read_count
resource_access = threading.Semaphore(1)  # To protect the shared resource
writer_done = threading.Event()  # Event to signal when the writer has written

# Reader count
read_count = 0
stop_event = threading.Event()  # Event to stop the program

# Reader function
def reader(reader_id):
    global read_count
    while not stop_event.is_set():
        writer_done.wait()

        mutex.acquire()
        read_count += 1
        if read_count == 1:
            resource_access.acquire()
        mutex.release()

        print(f"Reader {reader_id} is reading the value: {shared_resource}")
        time.sleep(random.uniform(0.1, 1))

        mutex.acquire()
        read_count -= 1
        if read_count == 0:
            resource_access.release()
        mutex.release()

        time.sleep(random.uniform(0.5, 2))

# Writer function
def writer(writer_id):
    global shared_resource
    while not stop_event.is_set():
        resource_access.acquire()

        shared_resource = random.randint(1, 100)
        print(f"Writer {writer_id} updated the value to: {shared_resource}")
        writer_done.set()
        time.sleep(random.uniform(0.5, 1))

        resource_access.release()

        time.sleep(random.uniform(1, 3))

# Main function
if _name_ == "_main_":
    readers = [threading.Thread(target=reader, args=(i,)) for i in range(3)]
    writers = [threading.Thread(target=writer, args=(i,)) for i in range(1)]

    for r in readers:
        r.start()
    for w in writers:
        w.start()

    # Stop threads after 10 seconds
    time.sleep(10)
    stop_event.set()

    for r in readers:
        r.join()
    for w in writers:
        w.join()
    print("Program terminated.")
